toolchain/check/import_cpp.cpp

```diff
@@ -1020,8 +1021,14 @@ static auto ImportNameDeclIntoScope(Context& context, SemIR::LocId loc_id,
                                     SemIR::NameId name_id,
                                     clang::NamedDecl* clang_decl)
     -> SemIR::InstId {
-  SemIR::InstId inst_id =
-      ImportNameDecl(context, loc_id, scope_id, name_id, clang_decl);
+  SemIR::InstId inst_id = SemIR::InstId::None;
+  SemIR::ClangDeclStore& clang_decls = context.sem_ir().clang_decls();
+  if (SemIR::ClangDeclId clang_decl_id = clang_decls.Lookup(clang_decl);
+      clang_decl_id.has_value()) {
+    inst_id = clang_decls.Get(clang_decl_id).inst_id;
+  } else {
+    inst_id = ImportNameDecl(context, loc_id, scope_id, name_id, clang_decl);
+  }
```

Comment:
Performing the check at the top level here will mean we do the same hash table lookup multiple times in some cases. For example, for type handling we will now look up a class type twice, once here and once when we reach `MapRecordType`. Also the map lookup will never succeed for a declaration of a kind that we don't store in the map, such as a typedef or using declaration, so it's a wasted lookup in those cases.

I think due to the above, it'd be more efficient to sink the `clang_decls` lookups into the individual decl handling functions. That'd mean that instead of calling`ImportNamespaceDecl` directly in `ImportNameDecl`, we should be calling `AsCarbonNamespace`, and we should add logic to do the `clang_decls` lookup to the `ImportFunctionDecl` code path.

Reply:
Done, though I think reusing `AsCarbonNamespace()`, which we use when we only have the namespace declaration but not its scope actually generates another lookup to search for its parent instruction.
Am I missing something?
I could add the scope as another (optional?) parameter so if we have it we can pass it. WDYT?

Reply:
ok, I'll leave it as is for now.

I would like to understand this better though.
If we're searching for a name in `Cpp.X`, and we find it by a Clang lookup for that name in `X`, we shouldn't necessarily add that name to `Cpp.X`?
IIUC, if we don't, the next time we will search for `Cpp.X`, we will have to do another name lookup in Clang.

Reply:
I believe the TODO was handled in #5821.
Adding inline namespace tests in #5826.